fluids: SmartSim Online Data-Driven SGS training

jrwrigh opened the pull request:
Todo:
- [x] Documentation
- [x] Send subgrid stresses (not just filtered velocity products)
    - [x] Add SGS to dd training input DM
    - [x] Add calculation of subgrid stress to training input calculation
    - [x] Verify DM Vec data placement (if true, remove IS and copying mechanics)
- [x] Add test


Depends on #1223

jrwrigh commented:
I currently don't have a test setup for this case. I could do one to test just the generation of the training data (so nothing about the communication with the database), but the current build framework I have requires building with SmartRedis for that. I think requiring building with SmartRedis makes sense, but not sure if we want to make a work around for that just for testing purposes.

jeremylt commented on examples/fluids/README.md:
```diff
@@ -8,6 +8,8 @@ The state variables are mass density, momentum density, and energy density.
 
 The main Navier-Stokes solver for libCEED is defined in [`navierstokes.c`](navierstokes.c) with different problem definitions according to the application of interest.
 
+## Build Instructions
```
Minor
```suggestion
## Build instructions
```

jeremylt commented on examples/fluids/README.md:
```diff
@@ -231,7 +239,7 @@ For the case of a square/cubic mesh, the list of face indices to be used with `-
 ### Boundary conditions
 
-Boundary conditions for compressible viscous flows are notoriously tricky. Here we offer some recommendations
+Boundary conditions for compressible viscous flows are notoriously tricky. Here we offer some recommendations.
```
```suggestion
Boundary conditions for compressible viscous flows are notoriously tricky.
Here we offer some recommendations.
```

jrwrigh commented on examples/fluids/problems/sgs_dd_training.c:
```diff
@@ -0,0 +1,379 @@
+    PetscCall(DMGetGlobalVector(sgs_dd_train->filtered_grad_velo_proj->dm, &FilteredVelocityGradient));
+    PetscCall(VelocityGradientProjectionApply(sgs_dd_train->filtered_grad_velo_proj, FilteredFields_loc, FilteredVelocityGradient));
+
+    // Compute Training Data
+    {
```
Note to self: Fix formatting

jrwrigh commented:
I hereby declare this PR officially good for review!

jrwrigh commented:
Alright, I've reviewed over this and rebased onto main. Good to review. @jedbrown @jeremylt

jeremylt commented on examples/fluids/include/smartsim.h:
```diff
@@ -0,0 +1,28 @@
+#if defined(__clang_analyzer__)
+void SmartRedisCall(SRError);
+#else
+#define SmartRedisCall(...)                                                                                                           \
```
For what its worth, I think PETSc would call this `PetscSmartRedisCall`

jeremylt commented on tests/junit_common.py:
```diff
@@ -312,6 +312,47 @@ def diff_cgns(test_cgns: Path, true_cgns: Path, tolerance: float = 1e-12) -> str
     return proc.stderr.decode('utf-8') + proc.stdout.decode('utf-8')
 
 
+def test_case_output_string(test_case: TestCase, spec: TestSpec, mode: RunMode,
```
This is a pretty big change to the common JUNIT code - have you tested that this core file can drop into Ratel with these changes? @zatkins-dev comments?

zatkins-dev replied:
I'll review/test this afternoon with Ratel

zatkins-dev replied:
This looks like a pretty clean refactor of the output string generation, I can't imagine it causing any issues. I'll double check but tentatively looks good

zatkins-dev replied:
LGTM! Ratel tests ran without issue locally. If this is finalized, I can make a matching MR in Ratel for the changes.

jrwrigh commented:
Thanks! I've changed the `SmartRedisCall` to `PetscSmartRedisCall` as Jeremy suggested. I've also rebased onto `main` just to make sure everything is working correctly CI-wise.
